File: Guias_praticos/Guia05/Java/ListaLigada.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package pilha;

public class ListaLigada {

    MemoriaLL m = null;
    final int NIL = Integer.MAX_VALUE;
    int head = NIL;
    
    public ListaLigada(int n){
        this.m = new MemoriaLL(n);
        this.head = NIL;
    }
    
    public void insert(int chave){
        int x = this.m.allocate_object();
        this.m.next[x] = this.head;
        this.m.key[x] = chave;
        if( this.head != NIL){
            this.m.prev[head] = x;
        }
        this.head = x;
        this.m.prev[x] = NIL;
    }
    
    public void delete(int x){
        if(this.m.prev[x] != NIL){
            this.m.next[this.m.prev[x]] = this.m.next[x];
        } else{
            this.head = this.m.next[x];
        }
        if(this.m.next[x] != NIL){
            this.m.prev[this.m.next[x]] = this.m.prev[x];
        }
        this.m.free_object(x);
    }
    
    public int search(int k){
        int x = this.head;
        while(x != NIL && this.m.key[x] != k){
            x = this.m.next[x];
        }
        return x;
    }
    
    @Override public String toString(){
        String s = this.m.toString() + "head -> " + this.head;
        return s;
    } 
}

File: Guias_praticos/Guia06/Python/HashTable/ListaLigada.py
from io import StringIO
import sys
from MemoriaLL import MemoriaLL

class ListaLigada:

    NIL = sys.maxsize
    m = None
    head = NIL

    def __getitem__(self, index):
        return self.m[index]

    def __setitem__(self, key, item):
        self.m[key] = item

    def __init__(self, n):
        self.m = MemoriaLL(n)
        self.head = self.NIL

    def insert(self, chave):
        x = self.m.allocate_object()
        self.m.nxt[x] = self.head
        self.m.key[x] = chave
        if self.head != self.NIL:
            self.m.prv[self.head] = x
        self.head = x
        self.m.prv[x] = self.NIL

    def delete(self, x):
        if self.m.prv[x] != self.NIL:
            self.m.nxt[self.m.prv[x]] = self.m.nxt[x]
        else:
            self.head = self.m.nxt[x]
        if self.m.nxt[x] != self.NIL:
            self.m.prv[self.m.nxt[x]] = self.m.prv[x]
        self.m.free_object(x)

    def search(self, k):
        x = self.head
        while x != self.NIL and self.m.key[x] != k:
            x = self.m.nxt[x]
        return x
            
    def toString(self):
        s = self.m.toString() + "head -> {}".format(self.head)
        return s
